crates/red_knot_test/src/lib.rs

```diff
@@ -0,0 +1,97 @@
+use ordermap::map::OrderMap;
+use red_knot_python_semantic::types::check_types;
+use ruff_db::files::system_path_to_file;
+use ruff_db::parsed::parsed_module;
+use ruff_db::system::{DbWithTestSystem, SystemPathBuf};
+use rustc_hash::FxHasher;
+use std::hash::BuildHasherDefault;
+use std::path::PathBuf;
+
+type FxOrderMap<K, V> = OrderMap<K, V, BuildHasherDefault<FxHasher>>;
+
+type Failures = FxOrderMap<SystemPathBuf, matcher::FailuresByLine>;
+
+mod assertion;
+mod db;
+mod diagnostic;
+mod matcher;
+mod parser;
+
+/// Run given file path as a markdown test suite.
+///
+/// Panic on test failure, and print failure details.
+#[allow(clippy::print_stdout)]
+pub fn run(path: &PathBuf) {
+    let relpath: PathBuf = path
+        .components()
+        .rev()
+        .take_while(|component| {
+            if let std::path::Component::Normal(os_str) = component {
+                os_str.to_str().is_some_and(|s| s != "mdtest")
+            } else {
+                true
+            }
+        })
+        .collect::<Vec<_>>()
+        .iter()
+        .rev()
+        .collect();
```

Comment:
That was my first version. It failed on Windows in CI, with no useful clues in the failure output. I would much rather maintain this code than debug why the `strip_prefix` version failed on Windows 😆 If you want to debug that and then replace this code, I have no problem with that. (Though I would rather strip the prefix all the way up to and including `mdtest`, as this code does, not just strip only the manifest dir.)

Reply:
Yeah, you would have to join the string with `mdtest` and strip the entire prefix. 

That it fails on windows could be if one of the two paths is absolute and the other is not. 

I suggest adding a TODO or comment explaining why we use this more complex (and not entirely safe) way of stripping a prefix

Reply:
I can add a comment. By "not entirely safe", do you mean in case there's an inner directory within `mdtest` also named `mdtest`?

I can also try absolutizing both paths and see if that fixes Windows CI on the `CARGO_MANIFEST_DIR` approach. But I think that will cause problems if we move the tests to a different crate. In general it feels ugly to me to rely on `CARGO_MANIFEST_DIR` in an exported library function.

Reply:
> I can add a comment. By "not entirely safe", do you mean in case there's an inner directory within mdtest also named mdtest? I could fix this, but it would make the code even more complex :)

Yeah. It's unlikely and not worth fixing. I would rather use strip prefix :P but that's what I had in mind

> In general it feels ugly to me to rely on CARGO_MANIFEST_DIR in an exported library function.

Using `CARGO_MANIFEST_DIR` with a fallback when it isn't set is relatively common. The only challenge is that it needs to run in the crate **calling** the test. 

The counter argument to the current solution is that assuming a test location also feels somewhat ugly for a library function ;)

Reply:
Good point!

I'll look at using strip prefix instead, on the caller side, and see if absolutizing helps Windows.

Reply:
Whoops, I forgot to link to the insta code https://github.com/mitsuhiko/insta/blob/ad7baa69bf5fcd54554c81aa97f90c53d1277a2a/insta/src/macros.rs#L27

Reply:
Looks like the Windows problem is that rstest calls `canonicalize` on each path, so we have to do the same before trying to strip the prefix, so that both paths have the UNC-whatever thingies in them.

Switched back to using `CARGO_MANIFEST_DIR`.